packages/lyra/src/lyra.ts

```diff
@@ -120,36 +163,54 @@ export class Lyra<TSchema extends PropertiesSchema = PropertiesSchema> {
     let i = 0;
     let j = 0;
 
+    const documentIdsFromNonSearchable = this.getDocumentIDsFromWhere({
+      boolIndices,
+      numericIndices,
+    });
+
     for (const term of tokens) {
       for (const index of indices) {
-        const documentIDs = await this.getDocumentIDsFromSearch({
+        const documentIDsFromText = await this.getDocumentIDsFromSearch({
           ...params,
-          index,
-          term,
-          exact,
+          index: index,
+          term: term,
+          exact: exact,
+          where: where,
         });
 
+        // To avoid duplicates we need a subtraction between added document
+        // And retrieved ones.
+        // We check if this is the first
+        // We use non-null assertion because setIntersection will return null only if both set are null
+        // We are sure that documentsIDs is not null, so we are safew
+        const documentIDs = setIntersection(
+          documentIDsFromText,
+          documentIdsFromNonSearchable
+        )!;
+
+        setSubtraction(documentIDs, documentAdded);
```

Comment:
@micheleriva even if this PR is not getting merged, I think that the set difference solves #48